tee-worker/litentry/core/identity-verification/src/web2/email/mailer/sendgrid_mailer.rs

```diff
@@ -0,0 +1,97 @@
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+use super::{Mail, Mailer};
+use http::header::{AUTHORIZATION, CONTENT_TYPE};
+use http_req::response::Headers;
+use itc_rest_client::{
+	error::Error as HttpError,
+	http_client::{HttpClient, SendWithCertificateVerification},
+	rest_client::RestClient,
+	RestPath, RestPost,
+};
+use serde::Serialize;
+use std::{string::String, vec, vec::Vec};
+use url::Url;
+
+#[derive(Serialize)]
+struct Personalization {
+	to: Vec<Email>,
+}
+
+#[derive(Serialize, Clone)]
+pub(crate) struct Email {
+	pub email: String,
+	pub name: Option<String>,
+}
+
+#[derive(Serialize)]
+pub(crate) struct Content {
+	content_type: String,
+	value: String,
+}
+
+#[derive(Serialize)]
+pub(crate) struct SendGridMail {
+	personalizations: Vec<Personalization>,
+	from: Email,
+	subject: String,
+	content: Vec<Content>,
+}
+
+impl SendGridMail {
+	pub fn new(from_email: String, mail: Mail) -> Self {
+		let content = vec![Content { content_type: String::from("text/html"), value: mail.body }];
+		let to = Email { email: mail.to, name: None };
+		let from = Email { email: from_email, name: Some(String::from("Litentry")) };
+		Self {
+			personalizations: vec![Personalization { to: vec![to] }],
+			from,
+			subject: mail.subject,
+			content,
+		}
+	}
+}
+
+impl RestPath<String> for SendGridMail {
+	fn get_path(path: String) -> Result<String, HttpError> {
+		Ok(path)
+	}
+}
+
+pub struct SendGridMailer {
+	client: RestClient<HttpClient<SendWithCertificateVerification>>,
+	from: String,
+}
+
+impl SendGridMailer {
+	pub fn new(api_key: String, from_email: String) -> Self {
+		let base_url = Url::parse("https://api.sendgrid.com/v3/mail/send").unwrap();
+		let authorization = std::format!("Bearer {}", api_key);
+
+		let mut headers = Headers::new();
+		headers.insert(AUTHORIZATION.as_str(), &authorization);
+		headers.insert(CONTENT_TYPE.as_str(), "application/json");
+
+		let http_client = HttpClient::new(
+			SendWithCertificateVerification::new(vec![]),
+			true,
+			None,
+			Some(headers),
+			None,
+		);
+
+		Self { client: RestClient::new(http_client, base_url), from: from_email }
+	}
+}
+
+impl Mailer for SendGridMailer {
+	fn send(&mut self, mail: Mail) -> Result<(), String> {
+		let sendgrid_mail = SendGridMail::new(self.from.clone(), mail);
+		self.client
+			.post(String::default(), &sendgrid_mail)
```

Comment:
This means sendgrid account owner knows what emails were linked and when ?

Reply:
This is an interesting point. The SendGrid account owner could know what emails requested the verification code in the last few days, but this is true for any service. We would have to setup and manage our own SMTP server if we want to prevent this, but even if we do that it would be very difficult to convince users that this is really the case. There has to be some level of trust for this solution. Another solution could be to add the option to verify with ZK proofs, so users can decide what to use. The tradeoff is that the UX is really bad